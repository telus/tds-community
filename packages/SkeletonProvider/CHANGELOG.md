# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="2.0.1"></a>

## [2.0.1](https://github.com/telus/tds-community/compare/@tds/community-skeleton@2.0.0...@tds/community-skeleton@2.0.1) (2018-10-22)

**Note:** Version bump only for package @tds/community-skeleton

<a name="2.0.0"></a>

# [2.0.0](https://github.com/telus/tds-community/compare/@tds/community-skeleton@1.0.0...@tds/community-skeleton@2.0.0) (2018-10-22)

### Bug Fixes

- **community-skeleton:** remove browser prefixes ([1a67f8d](https://github.com/telus/tds-community/commit/1a67f8d))

### Features

- **community-skeleton:** added animations, removed variants ([#115](https://github.com/telus/tds-community/issues/115)) ([e72073e](https://github.com/telus/tds-community/commit/e72073e))

### BREAKING CHANGES

- **community-skeleton:** removed property variant

omit
