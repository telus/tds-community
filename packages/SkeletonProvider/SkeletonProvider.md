### Wrapping TDS elements

When used as a container for other components, it enhances a sub-set of TDS core components by adding `skeleton` property.
The `skeleton` property will transform the child component into a skeleton. The `skeleton` property can be set to an object
with options to customize how the skeleton will appear.

| TDS Core    | skeleton options | type   | default   | Description                                       |
| ----------- | ---------------- | ------ | --------- | ------------------------------------------------- |
| **Text**    | lines            | number | 1         | Number of lines to appear as a skeleton           |
|             | characters       | number | undefined | Character width of skeleton (default is full row) |
| **Heading** | characters       | number | undefined | Character width of skeleton (default is full row) |
| **Image**   | none             | n/a    | n/a       | Images always appear as a circle in skeleton form |

### Example

The following is an example component consisting of several TDS core components. By wrapping it in `<Skeleton />` we can easily
provide a skeletonized state for this component.

```jsx
<SkeletonProvider show={true}>
  <Card variant="grey">
    <Box inline between={3}>
      <Box>
        <Image
          src="image-example.jpg"
          rounded="circle"
          width={140}
          height={140}
          alt="Image of co-workers collaborating"
          skeleton
        />
      </Box>
      <Box>
        <Heading skeleton={{ characters: 5 }} level="h1">
          Newton / Dilucca
        </Heading>
        <Text skeleton={{ lines: 2, characters: 20 }}>
          euro-pop assault table monofilament uplink saturation point industrial grade monofilament
          corrupted A.I. network.
        </Text>
      </Box>
    </Box>
  </Card>
</SkeletonProvider>
```

### Custom skeleton rendering

You can override or provide a custom skeleton to any element by passing a component to the `skeleton` property. For example,
the community contribuited `<Testimonial />` component can be skeletonized like this:

```jsx
<SkeletonProvider show={true}>
  <Testimonial
    testimonial={`“My office is where ever my customers are. TELUS helps me stay connected whether I'm making a sale or doing payroll.”`}
    image={
      <Image
        src="image-example.jpg"
        rounded="circle"
        width={60}
        height={60}
        alt="Image of co-workers collaborating"
      />
    }
    title="Dave Smith, Foreman"
    additionalInfo="Pinnacle Reforestation"
    skeleton={
      /* Place custom skeleton componenent here: */
      () => (
        <SkeletonProvider>
          <Card>
            <Text skeleton={{ characters: 15 }}>s</Text>
            <Box inline between={2}>
              <Image rounded="circle" width={60} height={60} src={''} alt="loading..." skeleton />
              <Text skeleton={{ characters: 10, lines: 2 }}>s</Text>
            </Box>
          </Card>
        </SkeletonProvider>
      )
    }
  />
</SkeletonProvider>
```
