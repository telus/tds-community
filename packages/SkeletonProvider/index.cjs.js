require('./dist/index.css')
const SkeletonProvider = require('./dist/index.cjs')

module.exports = SkeletonProvider
