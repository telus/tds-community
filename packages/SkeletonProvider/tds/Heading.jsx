import React from 'react'
import PropTypes from 'prop-types'
import Heading from '@tds/core-heading'
import Skeleton from '../../Skeleton/Skeleton'
import withSkeleton from '../withSkeleton'

const HeadingSkeleton = ({ skeleton, ...rest }) => {
  if (!skeleton) {
    return <Heading {...rest} />
  }

  const { level } = rest

  const levels = {
    h1: 'lg',
    h2: 'md',
    h3: 'sm',
    h4: 'xs',
  }

  let { characters } = skeleton

  if (skeleton === true) {
    characters = undefined
  }

  return (
    <div style={{ maxWidth: '100%', display: 'inline-block', overflow: 'hidden' }}>
      <Skeleton characters={characters} size={level ? levels[level] : 'md'} />
    </div>
  )
}

HeadingSkeleton.propTypes = {
  skeleton: PropTypes.oneOfType([
    PropTypes.shape({
      characters: PropTypes.number,
    }),
    PropTypes.bool,
  ]),
}

HeadingSkeleton.defaultProps = {
  skeleton: undefined,
}

export default withSkeleton(HeadingSkeleton)(Heading)
