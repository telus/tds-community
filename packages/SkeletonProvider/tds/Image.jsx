import React from 'react'
import PropTypes from 'prop-types'
import Image from '@tds/core-image'
import withSkeleton from '../withSkeleton'
import styles from '../Skeleton.scss'

const ImageSkeleton = ({ skeleton, ...rest }) => {
  if (!skeleton) {
    return <Image {...rest} />
  }

  const baseClass = styles['skeleton-image']
  const circularClass = styles.circular

  const cssWidth = `${rest.width * 0.8}px`
  const cssHeight = `${rest.height * 0.8}px`

  const classes = [baseClass, circularClass].join(' ')

  return <div className={classes} style={{ width: cssWidth, height: cssHeight }} />
}

ImageSkeleton.propTypes = {
  skeleton: PropTypes.bool,
}

ImageSkeleton.defaultProps = {
  skeleton: false,
}

export default withSkeleton(ImageSkeleton)(Image)
