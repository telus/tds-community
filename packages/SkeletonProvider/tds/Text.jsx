import React from 'react'
import PropTypes from 'prop-types'
import Text from '@tds/core-text'
import uuid from 'uuid/v1'
import Skeleton from '../../Skeleton/Skeleton'
import withSkeleton from '../withSkeleton'

const TextSkeleton = ({ skeleton, ...rest }) => {
  if (!skeleton) {
    return <Text {...rest} />
  }

  const skSizes = {
    large: 'md',
    base: 'xs',
    small: 'xs',
    medium: 'sm',
  }

  const { size } = skeleton

  let { lines, characters } = skeleton

  const skSize = skSizes[size] || 'xs'
  if (skeleton === true) {
    lines = 1
    characters = undefined
  }

  if (!lines || lines < 1) {
    lines = 1
  }

  if (lines === 1) {
    return <Skeleton size={skSize} characters={characters} />
  }

  return (
    <div style={{ width: '100%', display: 'inline-block', overflow: 'hidden' }}>
      {[...Array(lines)].map(() => (
        <div key={uuid()}>
          <Skeleton size={skSize} characters={characters} />
        </div>
      ))}
    </div>
  )
}

TextSkeleton.propTypes = {
  skeleton: PropTypes.oneOfType([
    PropTypes.shape({
      characters: PropTypes.number,
      lines: PropTypes.number,
    }),
    PropTypes.bool,
  ]),
}

TextSkeleton.defaultProps = {
  skeleton: undefined,
}

export default withSkeleton(TextSkeleton)(Text)
