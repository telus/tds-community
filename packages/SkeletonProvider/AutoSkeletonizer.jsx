import React from 'react'
import Text from '@tds/core-text'
import Heading from '@tds/core-heading'
import Image from '@tds/core-image'
import Skeletons from './tds'
import withSkeleton from './withSkeleton'

const AutoSkeletonizer = children => {
  const skeletonChildren = React.Children.map(children, childElement => {
    let child = childElement

    if (child.props && 'children' in child.props) {
      child = React.cloneElement(child, {
        children: AutoSkeletonizer(child.props.children),
      })
    }

    if (child.props && 'skeleton' in child.props && typeof child.props.skeleton === 'function') {
      const EnhancedChild = withSkeleton(child.props.skeleton)(() => child)
      return <EnhancedChild {...child.props} />
    }

    if (child.type === Text) {
      return <Skeletons.Text {...child.props} />
    }

    if (child.type === Heading) {
      return <Skeletons.Heading {...child.props} />
    }

    if (child.type === Image) {
      return <Skeletons.Image {...child.props} />
    }

    return child
  })

  return skeletonChildren
}

export default AutoSkeletonizer
