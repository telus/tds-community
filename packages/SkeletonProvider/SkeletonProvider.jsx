import React, { Fragment } from 'react'
import PropTypes from 'prop-types'
import { withContext } from 'recompose'
import AutoSkeletonizer from './AutoSkeletonizer'

const SkeletonProvider = withContext({ show: PropTypes.bool }, props => ({ show: props.show }))(
  props => <Fragment>{AutoSkeletonizer(props.children)}</Fragment>
)

SkeletonProvider.propTypes = {
  show: PropTypes.bool,
}
SkeletonProvider.defaultProps = {
  show: true,
}

export default SkeletonProvider
