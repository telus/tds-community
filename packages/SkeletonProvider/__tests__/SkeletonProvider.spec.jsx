import React from 'react'
import { shallow } from 'enzyme'
import Heading from '@tds/core-heading'
import Text from '@tds/core-text'
import Image from '@tds/core-image'
import SkeletonProvider from '../SkeletonProvider'

const doShallow = props => shallow(<SkeletonProvider {...props} />)

describe('SkeletonProvider', () => {
  it('should render children as skeletons', () => {
    const skeleton = doShallow({
      children: (
        <div>
          <Heading level="h1" skeleton>
            Hello
          </Heading>
          <Image width={60} height={60} corners="rounded" src="" alt="skeleton image" />
          <Text skeleton={{ characters: 4, lines: 2 }}>Hello</Text>
        </div>
      ),
    })
    expect(skeleton.render()).toMatchSnapshot()
  })

  it('should render children as is if the show prop == false', () => {
    const skeleton = doShallow({
      show: false,
      children: (
        <div>
          <Heading level="h1" skeleton>
            Hello
          </Heading>
          <Image width={60} height={60} corners="rounded" src="" alt="image" />
          <Text skeleton={{ characters: 4, lines: 2 }}>Hello</Text>
        </div>
      ),
    })
    expect(skeleton.render()).toMatchSnapshot()
  })
})
