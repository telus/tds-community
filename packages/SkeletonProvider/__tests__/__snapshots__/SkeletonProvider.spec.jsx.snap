// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkeletonProvider should render children as is if the show prop == false 1`] = `
<div>
  <h1
    class="TDS_Heading-modules__h1___1iaO2 TDS_Heading-modules__wordBreak___UVmND TDS_Spacing-modules__noSpacing___XPYDG TDS_Heading-modules__secondary___3skcz"
  >
    Hello
  </h1>
  <img
    alt="image"
    class="TDS_Image-modules__fluid___1qw3M"
    corners="rounded"
    height="60"
    src=""
    width="60"
  />
  <span
    class="TDS_Typography-modules__default___KlcpP TDS_Typography-modules__base___2nkPE TDS_Typography-modules__wordBreak___3OZx_ TDS_Typography-modules__baseFont___2Cfk4 TDS_Typography-modules__color___2CNH8"
  >
    Hello
  </span>
</div>
`;

exports[`SkeletonProvider should render children as skeletons 1`] = `
<div>
  <h1
    class="TDS_Heading-modules__h1___1iaO2 TDS_Heading-modules__wordBreak___UVmND TDS_Spacing-modules__noSpacing___XPYDG TDS_Heading-modules__secondary___3skcz"
  >
    Hello
  </h1>
  <img
    alt="skeleton image"
    class="TDS_Image-modules__fluid___1qw3M"
    corners="rounded"
    height="60"
    src=""
    width="60"
  />
  <span
    class="TDS_Typography-modules__default___KlcpP TDS_Typography-modules__base___2nkPE TDS_Typography-modules__wordBreak___3OZx_ TDS_Typography-modules__baseFont___2Cfk4 TDS_Typography-modules__color___2CNH8"
  >
    Hello
  </span>
</div>
`;
